src/clj/fluree/db/json_ld/iri.cljc

```diff
@@ -0,0 +1,201 @@
+
+
+(def default-namespace-codes
+  (map-invert default-namespaces))
+
+(def last-default-code 100)
+
+(def commit-namespaces
+  #{f-commit-256-ns})
+
+(def commit-namespace-codes
+  (into #{}
+        (map default-namespaces)
+        commit-namespaces))
+
+(defn decompose-by-char
+  [iri c limit]
+  (when-let [char-idx (some-> iri
+                              (str/last-index-of c)
+                              inc)]
+    (when (< char-idx limit)
+      (let [ns  (subs iri 0 char-idx)
+            nme (subs iri char-idx)]
+        [ns nme]))))
+
+(defn decompose
+  [iri]
+  (when iri
+    (let [iri*   (normalize iri)
+          length (count iri*)]
+      (or (decompose-by-char iri* \@ length)
+          (decompose-by-char iri* \# length)
+          (decompose-by-char iri* \? length)
+          (decompose-by-char iri* \/ length)
+          (decompose-by-char iri* \: length)
+          ["" iri*]))))
+
+(def name-code-xf
+  (comp (partition-all 8)
+        (map bytes/UTF8->long)))
+
+#?(:clj (defn name->codes
+          [nme]
+          (->> nme
+               bytes/string->UTF8
+               (into [] name-code-xf))))
+
+#?(:clj (defn codes->name
+          [nme-codes]
+          (->> nme-codes
+               (mapcat bytes/long->UTF8)
+               bytes/UTF8->string)))
+
+(defn ->sid
+  [ns-code nme]
+  (let [ns-int (int ns-code)
+        nme*   #?(:clj (-> nme name->codes long-array)
+                  :cljs nme)]
+    (SID. ns-int nme*)))
+
+(defn get-ns-code
+  [^SID sid]
+  #?(:clj (.getNamespaceCode sid)
+     :cljs (:namespace-code sid)))
+
+(defn get-namespace
+  ([sid]
+   (get-namespace sid default-namespace-codes))
+  ([sid namespace-codes]
+   (let [ns-code (get-ns-code sid)]
+     (get namespace-codes ns-code))))
+
+(defn get-name
+  [^SID sid]
+  #?(:clj (->> sid .getNameCodes codes->name)
+     :cljs (:name sid)))
+
+(defn deserialize-sid
+  [[ns-code nme]]
+  (->sid ns-code nme))
+
+(def serialize-sid
+  (juxt get-ns-code get-name))
+
+#?(:clj (defmethod print-method SID [^SID sid ^java.io.Writer w]
+          (doto w
+            (.write "#SID ")
+            (.write (-> sid serialize-sid pr-str)))))
+
+#?(:clj (defmethod print-dup SID
+          [^SID sid ^java.io.Writer w]
+          (let [ns-code (get-ns-code sid)
+                nme     (get-name sid)]
+            (.write w (str "#=" `(->sid ~ns-code ~nme))))))
+
+(defn sid?
+  [x]
+  (instance? SID x))
+
+(def ^:const min-sid
+  (->sid 0 ""))
+
+(def ^:const max-sid
+  (->sid util/max-integer ""))
+
+(defn iri->sid
+  "Converts a string iri into a vector of long integer codes. The first code
+  corresponds to the iri's namespace, and the remaining codes correspond to the
+  iri's name split into 8-byte chunks"
+  ([iri]
+   (iri->sid iri default-namespaces))
+  ([iri namespaces]
+   (let [[ns nme] (decompose iri)]
+     (when-let [ns-code (get namespaces ns)]
+       (->sid ns-code nme)))))
+
+(defn next-namespace-code
```

Comment:
It looks like we still have some state in the sid creation process, it lives in the `Db` under `:namespaces`.